language/stdlib/modules/payment_router.move

```diff
@@ -0,0 +1,144 @@
+    // to this multi-currency holding.
+    resource struct PaymentRouterInfo {
+        // Whether the router account has exlusive withdrawal privileges to
+        // its routed accounts.
+        exclusive_withdrawals_only: bool,
+        allowed_addresses: vector<address>,
+    }
+
+    // Withdrawal capabilities for the added accounts, partitioned by their
+    // currency type.
+    resource struct AccountInfo<Token> {
+        child_accounts: vector<address>,
+    }
+
+    // A delegated withdrawal capability that is published under the child
+    // account and that allows the routing account withdraw from it (and
+    // the sender, depending on flags set in `PaymentRouterInfo`).
+    resource struct RoutedAccount<Token> {
+        router_account_addr: address,
+        withdrawal_cap: LibraAccount::WithdrawalCapability,
+    }
+
+    // Initialize the sending account with `exclusive_withdrawals_only`
+    // either turned on (true) or off (false).
+    public fun initialize(exclusive_withdrawals_only: bool) {
+        move_to_sender(PaymentRouterInfo {
+            exclusive_withdrawals_only,
+            allowed_addresses: Vector::empty(),
+        })
+    }
+
+    // Set whether exclusive access is held to the routed accounts
+    public fun set_exclusive_withdrawals(exclusive_withdrawals_only: bool)
+    acquires PaymentRouterInfo {
+        borrow_global_mut<PaymentRouterInfo>(Transaction::sender())
+            .exclusive_withdrawals_only = exclusive_withdrawals_only
+    }
+
+    // Return whether the router at `router_account_addr` only allows
+    // exlusive withdrawals
+    public fun exclusive_withdrawals_only(router_account_addr: address): bool
+    acquires PaymentRouterInfo {
+        borrow_global_mut<PaymentRouterInfo>(router_account_addr).exclusive_withdrawals_only
+    }
+
+    // Allow the account at `addr` to delegate its withdrawal_capability to us.
+    public fun allow_account_address(addr: address)
+    acquires PaymentRouterInfo {
+        let sender = Transaction::sender();
+        let router_info = borrow_global_mut<PaymentRouterInfo>(sender);
+        if (!Vector::contains(&router_info.allowed_addresses, &addr))
+            Vector::push_back(&mut router_info.allowed_addresses, addr);
+    }
+
+    // Allow routing of currencies of `Token` type.
+    public fun allow_currency<Token>() {
+        move_to_sender(AccountInfo<Token>{ child_accounts: Vector::empty() })
+    }
+
+    // Add the sending account of currency type `Token` to the router at
+    // `router_account_addr`.  The sending address must be in the
+    // `allowed_addresses` held under the router account.
+    public fun add_account_to<Token>(router_account_addr: address)
+    acquires PaymentRouterInfo, AccountInfo {
+        let sender = Transaction::sender();
+        let router_info = borrow_global_mut<PaymentRouterInfo>(router_account_addr);
+        let (has, index) = Vector::index_of(&router_info.allowed_addresses, &sender);
```

Comment:
I wonder if it might make more sense to introduce `Vector::remove`?  That would lead to the shortest code here.

Reply:
Good question. My concern with adding a `remove` based on the elem is that there are a number of different semantic ambiguities around this that would need to be ironed out first. In particular:
* Does it only remove the first occurence? Or remove all of the elements?
* Does this do a swap_remove, or an order-preserving remove.
For now I'd like the to keep the `index_of` followed by `remove_swap` for this reason, but I think these functionalities should be added to the Vector module at some point.

Reply:
Got it, that reasoning makes sense. Thanks!